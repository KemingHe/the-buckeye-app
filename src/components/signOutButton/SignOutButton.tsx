"use client";

import { ArrowRightStartOnRectangleIcon } from "@heroicons/react/24/solid";

// biome-ignore format: added alignment for clarity.
export interface SignOutButtonProps {
  signOutAction: () => void | Promise<void>;
}

export function SignOutButton({
  signOutAction
}: SignOutButtonProps): JSX.Element {
  return (
    <button
      aria-label="Sign out of your account"
      className="btn btn-accent btn-sm flex items-center justify-center"
      onClick={signOutAction}
      type="button"
    >
      Sign Out
      <ArrowRightStartOnRectangleIcon aria-hidden="true" className="size-4" />
    </button>
  );
}
