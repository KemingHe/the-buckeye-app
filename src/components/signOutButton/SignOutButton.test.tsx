import {
  type RenderResult,
  render,
  screen,
  waitFor
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { beforeEach, describe, expect, it, vi } from "vitest";

import { SignOutButton } from "./SignOutButton";

const mockSignOutAction = vi.fn();

describe("SignOutButton component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders the sign out button with the correct text content", () => {
    render(<SignOutButton signOutAction={mockSignOutAction} />);

    const elementByText: HTMLElement = screen.getByText("Sign Out");
    expect(elementByText).toBeVisible();

    const elementByRole: HTMLElement = screen.getByRole("button", {
      name: "Sign out of your account"
    });
    expect(elementByRole).toBeVisible();
  });

  it("calls the sign out action when the button is clicked", async () => {
    render(<SignOutButton signOutAction={mockSignOutAction} />);
    const button: HTMLElement = screen.getByRole("button", {
      name: "Sign out of your account"
    });
    await waitFor(() => {
      expect(mockSignOutAction).toHaveBeenCalledTimes(0);
    });

    userEvent.click(button);
    await waitFor(() => {
      expect(mockSignOutAction).toHaveBeenCalledTimes(1);
    });

    userEvent.click(button);
    await waitFor(() => {
      expect(mockSignOutAction).toHaveBeenCalledTimes(2);
    });
  });

  it("matches the snapshot", () => {
    const { asFragment }: RenderResult = render(
      <SignOutButton signOutAction={mockSignOutAction} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
