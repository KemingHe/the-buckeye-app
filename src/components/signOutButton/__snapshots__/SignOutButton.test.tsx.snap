// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignOutButton component > matches the snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="Sign out of your account"
    class="btn btn-accent btn-sm flex items-center justify-center"
    type="button"
  >
    Sign Out
    <svg
      aria-hidden="true"
      class="size-4"
      data-slot="icon"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M7.5 3.75A1.5 1.5 0 0 0 6 5.25v13.5a1.5 1.5 0 0 0 1.5 1.5h6a1.5 1.5 0 0 0 1.5-1.5V15a.75.75 0 0 1 1.5 0v3.75a3 3 0 0 1-3 3h-6a3 3 0 0 1-3-3V5.25a3 3 0 0 1 3-3h6a3 3 0 0 1 3 3V9A.75.75 0 0 1 15 9V5.25a1.5 1.5 0 0 0-1.5-1.5h-6Zm10.72 4.72a.75.75 0 0 1 1.06 0l3 3a.75.75 0 0 1 0 1.06l-3 3a.75.75 0 1 1-1.06-1.06l1.72-1.72H9a.75.75 0 0 1 0-1.5h10.94l-1.72-1.72a.75.75 0 0 1 0-1.06Z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;
